tests/components/withings/test_sensor.py

```diff
@@ -0,0 +1,767 @@
+        data_manager_mock.return_value = data_manager_instance
+        health_sensor_mock.return_value = MagicMock(spec=WithingsHealthSensor)
+
+        async_add_entities = MagicMock()
+        config_entry = ConfigEntry(
+            1,
+            'domain',
+            'title',
+            {
+                const.PROFILE: 'Person 1',
+                const.CREDENTIALS: {
+                    'access_token': 'my_access_token',
+                    'token_expiry': 'my_token_expiry',
+                    'token_type': 'my_token_type',
+                    'refresh_token': 'my_refresh_token',
+                    'user_id': 'my_user_id',
+                    'client_id': 'my_client_id',
+                    'consumer_secret': 'my_consumer_secret'
+                },
+            },
+            'source',
+            'connection_class'
+        )
+
+        await async_setup_entry(
+            hass,
+            config_entry,
+            async_add_entities
+        )
+
+        async_add_entities.assert_called_with(ANY, True)
+
+
+def test_create_sensor_entities_all(hass: HomeAssistantType):
+    """Test entity creation for all."""
+    data_manager = MagicMock(spec=WithingsDataManager)
+    hass.data[const.DOMAIN] = {
+        const.CONFIG: {}
+    }
+
+    entities = create_sensor_entities(hass, data_manager)
+    assert entities
+    assert len(entities) == len(WITHINGS_MEASUREMENTS_MAP)
+
+
+def test_create_sensor_entities_skip(hass: HomeAssistantType):
+    """Test entity creation skipped."""
+    data_manager = MagicMock(spec=WithingsDataManager)
+    hass.data[const.DOMAIN] = {
+        const.CONFIG: {
+            const.MEASURES: [
+                const.MEAS_BODY_TEMP_C
+            ]
+        }
+    }
+
+    entities = create_sensor_entities(hass, data_manager)
+    assert entities
+    assert len(entities) == 1
+    assert entities[0] == WithingsHealthSensor(
+        data_manager,
+        WITHINGS_MEASUREMENTS_MAP[const.MEAS_BODY_TEMP_C]
+    )
+
+
+@pytest.fixture(name='nokia_api')
+def noka_credentials_fixture():
+    """Provide nokia credentials."""
+    api = nokia.NokiaApi.__new__(nokia.NokiaApi)
+    api.get_credentials = MagicMock(
+        return_value=nokia.NokiaCredentials(
+            user_id='USER_ID'
+        )
+    )
+    api.get_measures = MagicMock()
+    api.get_sleep = MagicMock()
+    api.get_sleep_summary = MagicMock()
+
+    return api
+
+
+@pytest.fixture(name='data_manager_factory')
+def data_manager_factory_fixture(nokia_api: nokia.NokiaApi):
+    """Provide a data manager factory function."""
+    def factory():
+        """Provide data manager."""
+        return WithingsDataManager(
+            'person_1',
+            nokia_api
+        )
+
+    return factory
+
+
+def test_health_sensor_properties(
+        data_manager_factory: Callable[[], WithingsDataManager]
+):
+    """Test method."""
+    attribute = WITHINGS_MEASUREMENTS_MAP[const.MEAS_WEIGHT_KG]
+    sensor = WithingsHealthSensor(
```

Comment:
Don't assert things about the entity directly. Set up the component which should set up the platform. Use `async_setup_component`. Let the platform add an entity. Then from the core state machine, get the state of the entity and assert the state and the state attributes.

The state is the truth of the entity.

Writing the platform tests like this will make them more robust.

Reply:
These sensors don't get created until after the config flow and config entry hooks are fired. That whole process is covered in other tests. These tests focus primarily on confirming the unit conversion code is working properly.

Reply:
The point is that we shouldn't interact with the entity directly in tests about the platform. We should use the state as the truth of the entity.

Reply:
I don't understand what is being asked here. The test in question "test_health_sensor_properties" only tests the mapping of properties when the entity is created. It doesn't use any platform fixtures. I'm not sure how this applies when you say "shouldn't interact with the entity directly in tests about the platform".

Reply:
I'm asking you to do this instead:
> Set up the component which should set up the platform. Use async_setup_component. Let the platform add an entity. Then from the core state machine, get the state of the entity and assert the state and the state attributes.

Reply:
This test's purpose is to confirm the variables passed to the constructor are correctly provided as properties with the same name.

Using async_setup_component sounds like a lot of overhead for what I'm trying to test here.

Using async_setup_component means I have to supply a component config for withings and it's dependencies. From there withings' async_setup is called which I have to mock configs so the setup is forwarded to the sensor platform. Once there, over a dozen WithingsSensor instances are created.

I already have tests that cover these platform interactions.

Reply:
Then we can remove this test. We shouldn't write tests that enforce rigidity in the entity implementation. We should be able to refactor the platform without changing tests, as long as we don't make breaking changes. Breaking changes will occur if we change what state the entity reports.

Reply:
Make sure we test all the state and state attributes of the entity in some tests, as I've described above.

Reply:
That makes sense. Looking into it, I already have coverage for this. I'll just remove this unnecessary test.